File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.cpp
class Solution {
private:
    vector<vector<int>> graph;
    vector<int> cycle, dist;
    int answer = -1;
public:
    void dfs(int src, int c, int res, vector<int>& edges) {
        if (dist[src] != -1) {
            if (cycle[src] != c) return;
            answer = max(answer, res - dist[src]);
            return;
        }
        dist[src] = res;
        cycle[src] = c;
        if (edges[src] != -1) {
            dfs(edges[src], c, res + 1, edges);
        }
    }
    int longestCycle(vector<int>& edges) {
        int n = edges.size();
        graph.resize(n);
        cycle.resize(n, 0);
        dist.resize(n, -1);
        int cycle_count = 0;
        for (int i = 0; i < n; i++) {
            if (dist[i] == -1) {
                dfs(i, cycle_count, 0, edges);
                cycle_count++;
            }
        }
        return answer;
    }
};

File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.py
class Graph:
    def __init__(self, n):
        self.vertices = n
        self.graph = defaultdict(set)
        self.visited = [False] * n
        self.parent = [-1] * n

    def add_edge(self, u, v):
        self.graph[u].add(v)
    
    def dfs(self, src, visited, rec_stack, cycle):
        visited[src] = True
        rec_stack[src] = True
        cycle[0] += 1
        for dest in self.graph[src]:
            if not visited[dest]:
                if self.dfs(dest, visited, rec_stack, cycle):
                    return True
            elif rec_stack[dest]:
                return True
        rec_stack[src] = False
        cycle[0] -= 1
        return False
    
    def longest_cycle(self):
        maxlength = -1
        visited, rec_stack = [False] * self.vertices, [False] * self.vertices 
        for src in range(self.vertices):
            if not self.visited[src]:
                cycle = [0]
                if self.dfs(src, visited, rec_stack, cycle):
                    maxlength = max(maxlength, cycle[0])
        return maxlength

class Solution:
    def longestCycle(self, edges: List[int]) -> int:
        graph = Graph(len(edges))
        for i, v in enumerate(edges):
            if v != -1:
                graph.add_edge(v, i)
        return graph.longest_cycle()